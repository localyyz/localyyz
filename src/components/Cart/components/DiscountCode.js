import React from "react";
import { inject, observer } from "mobx-react/native";
import { View, TouchableOpacity, StyleSheet, TextInput } from "react-native";
import EntypoIcon from "react-native-vector-icons/Entypo";
import FontAwesomeIcon from "react-native-vector-icons/FontAwesome";
import * as Animatable from "react-native-animatable";
import PropTypes from "prop-types";
import {
  Colours,
  Sizes,
  Styles,
  DEFAULT_DISCOUNT_CODE
} from "localyyz/constants";
import { UppercasedText } from "localyyz/components";
import CartHeader from "./CartHeader";
import CartField from "./CartField";

@inject(stores => ({
  applyDiscountCode: stores.cartStore.applyDiscountCode
}))
@observer
export default class DiscountCode extends React.Component {
  static propTypes = {
    applyDiscountCode: PropTypes.func.isRequired
  };

  constructor(props) {
    super(props);
    this.state = {
      isVerified: false,
      isOpen: false,
      discountCode: "",
      newDiscountCode: ""
    };

    this.updateDiscountCode = this.updateDiscountCode.bind(this);
    this.verifyDiscountCode = this.verifyDiscountCode.bind(this);
    this.toggleDiscountForm = this.toggleDiscountForm.bind(this);
  }

  toggleDiscountForm = () => {
    this.setState({ isOpen: !this.state.isOpen });
  };

  updateDiscountCode = discountCode => {
    this.setState({ newDiscountCode: discountCode });
  };

  verifyDiscountCode = () => {
    console.log(this.state.newDiscountCode);
    if (
      this.state.newDiscountCode.length > 0
      && this.props.applyDiscountCode({
        discountCode: this.state.newDiscountCode
      })
    ) {
      this.setState({
        isVerified: true,
        isOpen: false,
        discountCode: this.state.newDiscountCode
      });
    } else {
      this.setState({
        isVerified: false,
        isOpen: true,
        discountCode: ""
      });
    }
  };

  get renderDiscountForm() {
    return (
      <View style={style.formContainer}>
        <Animatable.View
          animation="fadeInDown"
          duration={200}
          style={[Styles.Horizontal, Styles.EqualColumns]}>
          <CartField
            icon="attach-money"
            label="Discount Code"
            color={Colours.EmphasizedText}>
            <TextInput
              ref="discount_code"
              autoCorrect={true}
              autoCapitalize="characters"
              placeholder={DEFAULT_DISCOUNT_CODE}
              placeholderTextColor={Colours.SubduedText}
              style={Styles.Input}
              value={this.state.discountCode}
              onChangeText={discount => this.updateDiscountCode(discount)}
              onEndEditing={this.verifyDiscountCode}/>
          </CartField>
        </Animatable.View>
        <TouchableOpacity
          style={style.addDiscount}
          onPress={this.verifyDiscountCode}>
          <View style={Styles.RoundedButton}>
            <UppercasedText style={style.addButtonLabel}>
              apply code
            </UppercasedText>
          </View>
        </TouchableOpacity>
      </View>
    );
  }
  render() {
    return (
      <View>
        <TouchableOpacity onPress={() => this.toggleDiscountForm()}>
          <CartHeader
            ref="verifyDiscountIcon"
            title="Discount code"
            icon={
              this.state.isVerified ? (
                <FontAwesomeIcon
                  name="check-circle"
                  size={Sizes.IconButton / 2}
                  color={Colours.PositiveButton}
                  style={Styles.IconOffset}/>
              ) : (
                <EntypoIcon
                  name="dot-single"
                  size={Sizes.IconButton}
                  color={Colours.NegativeButton}/>
              )
            }>
            {this.state.isVerified
              ? this.state.discountCode
              : "no discount code applied"}
          </CartHeader>
        </TouchableOpacity>
        {this.state.isOpen ? this.renderDiscountForm : null}
      </View>
    );
  }
}

const style = StyleSheet.create({
  formContainer: {
    paddingVertical: Sizes.InnerFrame,
    backgroundColor: Colours.Foreground
  },
  addDiscount: {
    marginHorizontal: Sizes.InnerFrame,
    marginVertical: Sizes.InnerFrame,
    alignItems: "flex-start"
  },
  addButtonLabel: {
    ...Styles.Text,
    ...Styles.Emphasized,
    ...Styles.Alternate
  }
});
