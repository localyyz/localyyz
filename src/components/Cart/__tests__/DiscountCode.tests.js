import React from "react";
import DiscountCode from "../components/DiscountCode";
import Renderer from "react-test-renderer";

const props = {
  applyDiscountCode: ({ discountCode }) => {
    if (discountCode == "SAMPLEDISCOUNT") {
      return true;
    } else {
      return false;
    }
  },
  updateDiscountCodeInStore: jest.fn()
};
describe("DiscountCode", () => {
  it("DiscountCode: should be able to enter valid discount code", () => {
    let rendered = Renderer.create(
      <DiscountCode.wrappedComponent {...props} />
    ).getInstance();
    rendered.toggleDiscountForm();
    rendered.updateDiscountCode("SAMPLEDISCOUNT");
    expect(rendered.state.newDiscountCode).toBe("SAMPLEDISCOUNT");
    expect(rendered.state.isVerified).toBe(false);
    expect(rendered.state.isOpen).toBe(true);
    rendered.verifyDiscountCode();
    expect(rendered.state.discountCode).toBe("SAMPLEDISCOUNT");
    expect(rendered.state.isVerified).toBe(true);
    expect(rendered.state.isOpen).toBe(false);
    expect(props.updateDiscountCodeInStore).toHaveBeenCalledTimes(1);
    expect(rendered.refs.verifyDiscountIcon.props.icon.props.name).toBe(
      "check-circle"
    );
  });

  it("DiscountCode: should not be able enter discount codes where server returns false", () => {
    let rendered = Renderer.create(
      <DiscountCode.wrappedComponent {...props} />
    ).getInstance();
    rendered.toggleDiscountForm();
    rendered.updateDiscountCode("BADDISCOUNT");
    rendered.verifyDiscountCode();
    expect(rendered.state.discountCode).toBe("");
    expect(rendered.state.isVerified).toBe(false);
    expect(rendered.state.isOpen).toBe(true);
    expect(rendered.refs.verifyDiscountIcon.props.icon.props.name).toBe(
      "dot-single"
    );
  });

  it("DiscountCode: should be able to clear discount code", () => {
    let rendered = Renderer.create(
      <DiscountCode.wrappedComponent {...props} />
    ).getInstance();
    rendered.toggleDiscountForm();
    rendered.updateDiscountCode("SAMPLEDISCOUNT");
    expect(rendered.state.newDiscountCode).toBe("SAMPLEDISCOUNT");
    expect(rendered.state.isVerified).toBe(false);
    expect(rendered.state.isOpen).toBe(true);
    rendered.verifyDiscountCode();
    expect(rendered.state.discountCode).toBe("SAMPLEDISCOUNT");
    expect(rendered.state.isVerified).toBe(true);
    expect(rendered.state.isOpen).toBe(false);
    expect(rendered.refs.verifyDiscountIcon.props.icon.props.name).toBe(
      "check-circle"
    );
    rendered.toggleDiscountForm();
    rendered.updateDiscountCode("");
    rendered.verifyDiscountCode();
    expect(rendered.state.discountCode).toBe("");
    expect(rendered.state.isVerified).toBe(false);
    expect(rendered.state.isOpen).toBe(true);
    expect(rendered.refs.verifyDiscountIcon.props.icon.props.name).toBe(
      "dot-single"
    );
  });

  it("Discount Code: snapshot regression test", () => {
    let rendered = Renderer.create(
      <DiscountCode.wrappedComponent {...props} />
    ).toJSON();
    expect(rendered).toMatchSnapshot();
  });
});
