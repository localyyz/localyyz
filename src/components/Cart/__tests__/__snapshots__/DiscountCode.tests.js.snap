// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DiscountCode Discount Code: snapshot regression test 1`] = `
<View>
  <View
    accessible={true}
    isTVSelectable={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "flex-start",
          "marginHorizontal": 15,
          "marginVertical": 15,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "backgroundColor": "rgba(255, 255, 255, 0)",
            "color": "#000000",
            "fontSize": 16,
            "fontWeight": "600",
          }
        }
      >
        Discount code
      </Text>
      <View
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "flexDirection": "row",
            "justifyContent": "flex-start",
            "marginHorizontal": 7.5,
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "backgroundColor": "rgba(255, 255, 255, 0)",
              "color": "rgb(30, 30, 30)",
              "fontSize": 16,
              "fontWeight": "300",
            }
          }
        >
          no discount code applied
        </Text>
      </View>
      <Text
        accessible={true}
        allowFontScaling={false}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "color": "#FF0000",
              "fontSize": 25,
            },
            undefined,
            Object {
              "fontFamily": "Entypo",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
          ]
        }
      >
        ï†€
      </Text>
    </View>
  </View>
</View>
`;
